pkg/utils/probe_helper.go

```diff
@@ -56,30 +57,66 @@ func (d *DeploymentProber) ProbeFields() []agent.ProbeField {
 	return probeFields
 }
 
-func DeploymentAvailabilityHealthCheck(identifier workapiv1.ResourceIdentifier, result workapiv1.StatusFeedbackResult) error {
-	if identifier.Resource != "deployments" {
+func DeploymentAvailabilityHealthCheck(identifier workapiv1.ResourceIdentifier,
+	result workapiv1.StatusFeedbackResult) error {
+	return WorkloadAvailabilityHealthCheck(identifier, result)
+}
+
+func WorkloadAvailabilityHealthCheck(identifier workapiv1.ResourceIdentifier,
+	result workapiv1.StatusFeedbackResult) error {
+	// only support deployments and daemonsets for now
+	if identifier.Resource != "deployments" && identifier.Resource != "daemonsets" {
 		return fmt.Errorf("unsupported resource type %s", identifier.Resource)
 	}
-	if identifier.Group != "apps" {
+	if identifier.Group != appsv1.GroupName {
 		return fmt.Errorf("unsupported resource group %s", identifier.Group)
 	}
 
 	if len(result.Values) == 0 {
-		return fmt.Errorf("no values are probed for deployment %s/%s", identifier.Namespace, identifier.Name)
+		return fmt.Errorf("no values are probed for %s %s/%s",
+			identifier.Resource, identifier.Namespace, identifier.Name)
 	}
+
+	readyReplicas := -1
+	desiredNumberReplicas := -1
 	for _, value := range result.Values {
-		if value.Name != "ReadyReplicas" {
-			continue
+		// for deployment
+		if value.Name == "ReadyReplicas" {
+			readyReplicas = int(*value.Value.Integer)
+		}
+		if value.Name == "Replicas" {
+			desiredNumberReplicas = int(*value.Value.Integer)
 		}
 
-		if *value.Value.Integer >= 1 {
-			return nil
+		// for daemonset
+		if value.Name == "NumberReady" {
+			readyReplicas = int(*value.Value.Integer)
+		}
+		if value.Name == "DesiredNumberScheduled" {
+			desiredNumberReplicas = int(*value.Value.Integer)
 		}
+	}
 
-		return fmt.Errorf("readyReplica is %d for deployment %s/%s",
-			*value.Value.Integer, identifier.Namespace, identifier.Name)
+	if readyReplicas == -1 {
+		return fmt.Errorf("readyReplica is not probed")
+	}
+	if desiredNumberReplicas == -1 {
+		return fmt.Errorf("desiredNumberReplicas is not probed")
 	}
-	return fmt.Errorf("readyReplica is not probed")
+
+	switch identifier.Resource {
+	case "deployments":
+		if desiredNumberReplicas == 0 || readyReplicas >= 1 {
+			return nil
+		}
+	case "daemonsets":
+		if readyReplicas == desiredNumberReplicas && readyReplicas > -1 {
```

Comment:
@qiujian16 @zhiweiyin318  Do you think this is reasonable for daemonsets?

Reply:
If a cluster has no nodes, when deploying a daemonset, it will be like:
```
╰─$ oc get ds nginx-deployment
NAME               DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
nginx-deployment   0         0         0       0            0           <none>          18s
```
should we treat it as health or not?

Reply:
should be healthy in this case.